Add docs (#53)
datajoint/element-deeplabcut: docs/src/concepts.md

  design](https://moser-pipelines.readthedocs.io/en/latest/imaging/dlc.html)

- Mesoscale Activity Project (Janelia Research Campus/Baylor College of Medicine/New
  York University)

- Hui-Chen Lu Lab (Indiana University)

- Tobias Rose Lab (University of Bonn)

- James Cotton Lab (Northwestern University)

## Pipeline Development

Development of the Element began with an [open source
repository](https://github.com/MMathisLab/DataJoint_Demo_DeepLabCut) shared by the
Mathis team. We further identified common needs across our respective partnerships to
offer the following features for single-camera 2D models:

- Training data and parameter management
- Launching model training and automatic model evaluation
- Model metadata management
- Launching inference video analysis and capturing pose estimation output

The workflow handles training data as file sets stored within DLC's project directory.
Parameters of the configuration file are captured and preserved. Model evaluation
permits direct model comparison, and, when combined with upstream Elements, Element
DeepLabCut can be used to generate pose estimation information for each session.

> 

Reviewer: Please add a section describing the high-level concepts.

> 

Author: Sure. I can add some notes regarding NWB. What did you have in mind for publishing?

> 

Reviewer: Just a brief note about DANDI and the function we have for automated upload.

# Concepts

## Pose Estimation in Neurophysiology

> <!-- Info from previous BACKGROUND doc -->

Reviewer: ```suggestion
```

[DeepFly3D](https://github.com/NeLy-EPFL/DeepFly3D)) or uniquely 2D (e.g.,
[DeepPoseKit](https://github.com/jgraving/DeepPoseKit)), DLC highlights a diversity of
use-cases via a [Model Zoo](http://www.mackenziemathislab.org/dlc-modelzoo). Even
compared to tools with similar functionality (e.g.,
[SLEAP](https://github.com/murthylab/sleap) and
[dannce](https://github.com/spoonsso/dannce)), DLC has more users, as measured by either
GitHub forks or more citations (1600 vs. 900). DLC's trajectory toward an industry
standard is attributable to [continued
funding](http://www.mackenziemathislab.org/deeplabcutblog/2020/11/18/czidlc), [extensive
documentation](https://deeplabcut.github.io/DeepLabCut/docs/intro.html) and both
creator- and peer-support. Other comperable tools include
[mmpose](https://github.com/open-mmlab/mmpose),
[idtracker.ai]([idtracker.ai](https://idtrackerai.readthedocs.io/en/latest/)),
[TREBA](https://github.com/neuroethology/TREBA),
[B-KinD](https://github.com/neuroethology/BKinD),
[VAME](https://github.com/LINCellularNeuroscience/VAME), and
[MARS](https://github.com/neuroethology/MARS).

## Element Architecture

> Each node in the following diagram represents the analysis code in the Element DeepLabCut and corresponding table in the database.

Reviewer: ```suggestion
Each node in the following diagram represents the analysis code in the workflow for Element DeepLabCut and corresponding table in the database.  Within the workflow, Element DeepLabCut connects to upstream Elements including Lab, Animal, and Session.
```

![element-deeplabcut diagram](https://raw.githubusercontent.com/datajoint/element-deeplabcut/main/images/diagram_dlc.svg)

> - Upstream: 
> 
>     - Element DeepLabCut connects to a a ***Session*** table, which is modeled in our 
>     [workflow pipeline](https://github.com/datajoint/workflow-deeplabcut/blob/main/workflow_deeplabcut/pipeline.py). 
>     
>     - Although not requried, most choose to connect ***Session*** to a ***Subject** 
>     table for managing research subjects.

Reviewer: ```suggestion
### `lab` schema

| Table | Description |
| --- | --- |
| Device | Camera metadata |

### `subject` schema
- Although not required, most choose to connect the `Session` table to a `Subject` table.

| Table | Description |
| --- | --- |
| Subject | Basic information of the research subject |

### `session` schema

| Table | Description |
| --- | --- |
| Session | Unique experimental session identifier |
```

        Pairings of video recordings with models to be use for pose estimation.
    
    + ***PoseEstimation***: Results of pose estimation using a given model. 

## Key Partnerships

[Mackenzie Mathis](http://www.mackenziemathislab.org/) (Swiss Federal Institute of Technology Lausanne) is both a lead
developer of DLC and a key advisor on DataJoint open source development as a member of
the [Scientific Steering Committee](datajoint.com/docs/elements/management/governance).

DataJoint is also partnered with a number of groups who use DLC as part of broader
workflows. In these collaborations, members of the DataJoint team have interviewed
researchers to understand their needs in experiment workflow, pipeline design, and
interfaces.

These teams include:

- Moser Group (Norwegian University of Science and Technology) - see [pipeline
  design](https://moser-pipelines.readthedocs.io/en/latest/imaging/dlc.html)

- Mesoscale Activity Project (Janelia Research Campus/Baylor College of Medicine/New
  York University)

- Hui-Chen Lu Lab (Indiana University)

- Tobias Rose Lab (University of Bonn)

- James Cotton Lab (Northwestern University)

> ## Pipeline Development

Reviewer: ```suggestion
## Element Development
```

DataJoint is also partnered with a number of groups who use DLC as part of broader
workflows. In these collaborations, members of the DataJoint team have interviewed
the scientists to understand their needs in experimental setup, pipeline design, and
interfaces.

These teams include:

- Moser Group (Norwegian University of Science and Technology) - see [pipeline
  design](https://moser-pipelines.readthedocs.io/en/latest/imaging/dlc.html)

- Mesoscale Activity Project (Janelia Research Campus/Baylor College of Medicine/New
  York University)

- Hui-Chen Lu Lab (Indiana University)

- Tobias Rose Lab (University of Bonn)

- James Cotton Lab (Northwestern University)

## Element Architecture

Each node in the following diagram represents the analysis code in the workflow and the corresponding tables in the database.  Within the workflow, Element DeepLabCut connects to upstream Elements including Lab, Animal, and Session.  For more detailed documentation on each table, see the API docs for the respective schemas.

![element-deeplabcut diagram](https://raw.githubusercontent.com/datajoint/element-deeplabcut/main/images/diagram_dlc.svg)

> ### `lab` schema ([API docs](https://datajoint.com/docs/workflow-deeplabcut/api/workflow_deeplabcut/pipeline/#workflow_deeplabcut.pipeline.Device))

Reviewer: ```suggestion
### `lab` schema ([API docs](./api/workflow_deeplabcut/pipeline/#workflow_deeplabcut.pipeline.Device))
```

| Table | Description |
| --- | --- |
| Device | Camera metadata |

> ### `subject` schema ([API docs](https://datajoint.com/docs/element-animal/api/element_animal/subject))

Reviewer: ```suggestion
### `subject` schema ([API docs](../element-animal/api/element_animal/subject))
```

- Although not required, most choose to connect the `Session` table to a `Subject` table.

| Table | Description |
| --- | --- |
| Subject | Basic information of the research subject |

> ### `session` schema ([API docs](https://datajoint.com/docs/element-session/api/element_session/session_with_datetime))

Reviewer: ```suggestion
### `session` schema ([API docs](../element-session/api/element_session/session_with_datetime))

| Table | Description |
| --- | --- |
| Session | Unique experimental session identifier |

> ### `train` schema ([API docs](/api/element_deeplabcut/train))

Reviewer: ```suggestion
### `train` schema ([API docs](./api/element_deeplabcut/train))
```

- Optional tables related to model training.

| Table | Description |
| --- | --- |
| VideoSet | Set of files corresponding to a training dataset. |
| TrainingParamSet | A collection of model training parameters, represented by an index. |
| TrainingTask | A set of tasks specifying model training methods. |
| ModelTraining | A record of training iterations launched by `TrainingTask`. |

> ### `model` schema ([API](/api/element_deeplabcut/model))

Reviewer: ```suggestion
### `model` schema ([API](./api/element_deeplabcut/model))
```

# Concepts

## Pose Estimation in Neurophysiology

> Neurophysiology is concerned not only with the inner workings of the brain, but the
> relationship between neural firings and environmental stimuli, natural behavior, or
> inferred cognitive states. One approach is to present a stationary subject with planned
> stimuli. More naturalistic paradigms permit spontaneous behavior, and add some mechanism
> measure subject responses. Through pose estimation, we capture the richness of a
> subject's natural behavior, which can then be paired with neuronal recordings.

Reviewer: ```suggestion
Studying the inner workings of the brain requires understanding the relationship between neural activity and environmental stimuli, natural behavior, or inferred cognitive states. Pose estimation is a computer vision method to track the position, and thereby behavior, of the subject over the course of an experiment, which can then be paired with neuronal recordings to answer scientific questions about the brain.
```

Previous pose estimation methods required reflective markers placed on a subject, as
well as multiple expensive high-frame-rate infrared cameras to triangulate position
within a limited field. Recent advancements in machine learning have facilitated
dramatic advancements in capturing pose data with a video camera alone. In particular,
[DeepLabCut](http://deeplabcut.org/) (DLC) facilitates the use of pre-trained machine 
learning models for 2-D and
3-D non-invasive markerless pose estimation. 

By training a model, DLC takes a machine learning process in its current state
and refining it to specialize in the video data collected in our experiment. This 
process works best when the camera has a consistent field of view across sessions, and
when generating pose estimation inferences on consistent data sets (e.g., subjects with 
similar characteristics). DLC saves models at various checkpoints and recommends
training for approximately 200,000 iterations. When DLC generates pose estimation 
inferences, this model is applied to novel videos, generating predictions of where the

> same set of training points would be in this new situation.

Reviewer: Please rephrase to improve clarity.  I am not sure of the purpose of this paragraph.
